File: solutions/dart/isogram/1/lib/isogram.dart
// Eternally offered to Shree DR.MDD

final REG_ALPHA = RegExp(r'\p{Alphabetic}', caseSensitive: false, unicode: true);

extension on String {
  bool isAlphabet() => REG_ALPHA.hasMatch(this);
}

class Isogram {
  bool isIsogram(String input) {
    final tracker = <String>{};
    for (final unit in input.toLowerCase().split('')) {
      if (unit.isAlphabet()) {
        if (!tracker.add(unit)) return false;
      }
    }
    return true;
  }
}

File: solutions/go/isogram/1/isogram.go
// Shree DR.MDD
package isogram

func IsIsogram(input string) bool {
	var mask uint32
	for _, r := range input {
		if !isAlpha(r) {
			continue
		}
		bit := letterMask(r)
		if mask&bit != 0 {
			return false
		}
		mask |= bit
	}
	return true
}

func isAlpha(r rune) bool {
	return (r >= 'A' && r <= 'Z') || (r >= 'a' && r <= 'z')
}

func letterMask(r rune) uint32 {
	if r >= 'a' && r <= 'z' {
		return 1 << uint(r-'a')
	}
	return 1 << uint(r-'A')
}
